### ISSUE TITLE
Using relation_aggregate() in labeling or symbology with large data is extremely slow

### ISSUE THREAD

**Issue by JuhoErvasti** [qgis/QGIS#51487]
### What is the bug or the crash?

Referring to a previous issue [#35896](https://github.com/qgis/QGIS/issues/35896) labeling with the relation_aggregate() function is inefficient as it makes separate requests to the referencing layer for each feature to be labeled.

### Steps to reproduce the issue

Utilizing the same database as in issue #47797 we have a testbed and environment to test this issue. Please follow these instructions:

**1. Establish a new PostGIS in connection in QGIS with following parameters**:

PostGIS instance IP: 18.195.144.58
Port: 5432
User: readonly
Password: readonly

**2. Open the QGIS project named qgis-project-relationaggr in the database**

- You can find it from qgisdbtest schema
- The project contain tables 'buildings' and 'land_parcel'. The tables have randomly generated data.

**3. Toggle the visibility of the _buildings_ or _land_parcel_ layer and see how long it takes for QGIS to load the labels and symbols**

- Additionally: Once a layer has been loaded once you can reproduce the issue again by toggling the visibility of a symbol off and back on

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
</head><body>

QGIS version | 3.22.13-Białowieża | QGIS code revision | 02ff117c
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.9.5
GDAL/OGR version | 3.6.0
PROJ version | 9.1.0
EPSG Registry database version | v10.074 (2022-08-01)
GEOS version | 3.11.1-CAPI-1.17.1
SQLite version | 3.39.4
PDAL version | 2.4.3
PostgreSQL client version | 14.3
SpatiaLite version | 5.0.1
QWT version | 6.1.6
QScintilla2 version | 2.13.1
OS version | Windows 10 Version 2009
 | | | 
Active Python plugins
db_manager | 0.1.20
grassprovider | 2.12.99
MetaSearch | 0.3.5
processing | 2.12.99

</body></html>QGIS version
3.22.13-Białowieża
QGIS code revision
[02ff117c](https://github.com/qgis/QGIS/commit/02ff117c)
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.6.0
PROJ version
9.1.0
EPSG Registry database version
v10.074 (2022-08-01)
GEOS version
3.11.1-CAPI-1.17.1
SQLite version
3.39.4
PDAL version
2.4.3
PostgreSQL client version
14.3
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009

Active Python plugins
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.5
processing
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by roya0045** [comment]
How is this a bug?

**Comment by troopa81** [comment]
### Problem

`relation_aggregate` method is called for every feature we want to label, and each time, we open a Postgres cursor, execute the request which look like the following

```sql
SELECT "id_contact"::text,"id_building"::text FROM "qgisdbtest"."building_owners" WHERE ("id_building" = 351816);
```

Then we get the result and close the cursor. This whole process takes (on my machine) approximatively 75ms.

This is because we repeat this request for each feature, that it ends up very slow. One could argue that we just have to make only one request for every feature we need to draw, but we know the feature that we need to draw only when we draw it. We don't load every feature in memory to not crash memory and we draw the feature just after reading it from database.

### Solution 

I tried to think of a solution, but it always ended up by an ugly complicated stuff very difficult to maintain. 

More importantly, I think it's a bad idea because IMHO, it's the database responsibility to correctly present the data. Regarding your use case, creat a view (or materialized one if you want even better performance) this way :

```sql
create view building_with_owners as select b.*, count(b.id) from qgisdbtest.building b JOIN qgisdbtest.building_owners o on o.id_building = b.id GROUP BY b.id;
```

Copy/Paste building style, change the category value with `count` and that's it, rendering is instantaneous.

Even if someone come up with a complicated scheme to improve things on QGIS side, he would NEVER manage to do something better.


**TL;DR** I think `relation_aggregate` is to be used on feature form when you display only one feature at a time. It's a Won't Fix to me.

**Comment by meyerlor** [comment]
Hey @troopa81 sorry to gravedig this thread!
I'm struggeling since quite some time (https://github.com/qgis/QGIS/issues/61804 ) with the same issue reported here (I found this issue just now) - @nyalldawson suggested some improvements to my expressions, but they did not improve the situation. My issue is not precicely with labeling, but with feature styles based on their children, the root cause is the same though.

My setup is a tree cadastre which i need to work with from: QGIS, Lizmap (=QGIS-server) and QField. My parent layer (trees) has ~800 features, the child layer (inspections) ~9000. It's hosted on PostGres, ~20ms RTT, the symbology categorises each tree by whether its last inspection is overdue, via `relation_aggregate()` - makes for my specific project ~166s (!) for a full render.

I did take your advice and built the view - in QGIS Desktop/Lizmap this is an okayish workaround, rendering the symbology is instant. It' wont work for QField though as it does not really support views. The largest draw back of the workaround imho is though, that a view can obviously not be edited. So i need to load the original "tree" layer, make it's symbology invisible but clickable to be able to work with the data.

Creating a solution for/in QGIS would probably not make rendering faster then a (materialized) view, but still would be much better than the current state PLUS the possibility to edit/update features. Also worth noting: this is not a "large dataset" problem as the title here might suggest - my 800 trees layer is not large.

---

I spent some time in the code (thanks, @claude ) to understand where the 166s actually go, and it came up with a concrete approach:

**Where the time goes.** Each `relation_aggregate()` call makes its own feature request, and on the Postgres provider every feature request is a full cursor lifecycle, not a single query (`QgsPostgresConn::openCursor` / `closeCursor`, `qgspostgresconn.cpp:1563-1591`):

```
BEGIN READ ONLY; DECLARE qgis_NNNN BINARY CURSOR FOR SELECT ...   16 ms
FETCH FORWARD 2000 FROM qgis_NNNN                                 16 ms
CLOSE qgis_NNNN; COMMIT                                           33 ms
                                                            ─────────
                                                                  65 ms
```

So the transaction bracket alone costs about as much as the query. Server side execution of one of these is under 1ms - almost everything is round trips.

**@roya0045 asked earlier whether these results are cached** - they are, but the cache key is built from `parameters.filter`, which contains the current feature's foreign key (`qgsexpressionfunction.cpp:1023`). So the key is different for every feature and the cache never hits across features. It only grows.

**Suggestion: take the parent feature's key out of the cache key.** Then the entry is shared by all features, and the cache becomes a lookup table instead of a log:

```
NOW                                       PROPOSED

tree 1   -> SELECT ... WHERE fk='a'       tree 1   -> miss -> SELECT fk, agg(x)
tree 2   -> SELECT ... WHERE fk='b'                           GROUP BY fk
tree 3   -> SELECT ... WHERE fk='c'                           -> map{a:.., b:.., c:..}
  :                                       tree 2   -> map[b]
tree 800 -> SELECT ... WHERE fk='z'       tree 3   -> map[c]
                                            :
                                          tree 800 -> map[z]

~800 cursor lifecycles                    1 cursor lifecycle
```

Filled lazily on the first miss: one pass over the child layer, reading only the FK column plus whatever the sub expression references, grouped into a hash in memory. No SQL and no `GROUP BY` pushdown needed, so it works for GeoPackage and WFS too - which is the part that matters for QField.

Why I think this is a reasonable shape:

- **No cache invalidation problem.** The map is stored in the `QgsExpressionContext`, which lives for exactly one layer render (`QgsVectorLayerRenderer` reuses one context and calls `setFeature()` per feature). It is created and thrown away inside the render, so there is nothing to invalidate and no risk of stale symbology. I originally thought this needs a cache that is renewed when a child feature changes - it does not, and that makes it much less risky.
- **Memory is not worse than today.** It stores one aggregated value per distinct FK value, never the child features themselves. Today's cache already stores one entry per parent feature for the whole render, so in practice this is usually smaller.
- **Narrow envelope.** Only single column FK, no `order_by`, sub expression referencing only child layer fields, and a size limit on the child layer. Everything else keeps exactly the current code path, so the change is easy to review and low risk.
- **Adaptive, so forms stay untouched.** Count the evaluations and pay per feature for the first ~10, then switch to batch mode. Identify and attribute forms never reach that threshold and behave exactly as now, a full render crosses it immediately. This also avoids the case where batching would lose, e.g. only a handful of features visible on screen.

I do not expect this to beat a materialized view, and I agree that the database is the better place when you can use it. But the gap between batched and view is milliseconds, while the gap to the current state is minutes - and unlike the view it keeps the layer editable and works offline in QField.

@troopa81 would you reconsider your "Won't fix" or do you still think this is not to be solved on QGIS side? I can work on a PR if you think that it's worth it. Would be delighted to hear from you in any case!

**Comment by troopa81** [comment]
> I did take your advice and built the view - in QGIS Desktop/Lizmap this is an okayish workaround, rendering the symbology is instant. 

I think it's not just an "okayish workaround", I think this is the appropriate straightforward way to go.

> It' wont work for QField though as it does not really support views. 

I don't know much about QField so I don't know about this limitation.

> The largest draw back of the workaround imho is though, that a view can obviously not be edited. 

a materialized one no, but a classic one yes.

> @troopa81 would you reconsider your "Won't fix" or do you still think this is not to be solved on QGIS side? I can work on a PR if you think that it's worth it. Would be delighted to hear from you in any case!

I take a real quick glance of your AI generated proposal and not sure to understand exactly what's proposed and the implication on the code.

This is a complicated matter and it requires funded time to dig into this before taking a decision.

**Comment by meyerlor** [comment]
thank you for taking the time to get back on this thread!



>     The largest draw back of the workaround imho is though, that a view can obviously not be edited.
> 
> a materialized one no, but a classic one yes.

You are right, simple views are editable but I need an aggregated view for the use case here, and those are not editable as stated in the [Postgres docs](https://www.postgresql.org/docs/current/sql-createview.html):

> Updatable Views
> 
> Simple views are automatically updatable: the system will allow INSERT, UPDATE, DELETE, and MERGE statements to be used on the view in the same way as on a regular table. A view is automatically updatable if it satisfies all of the following conditions:
> 
>     The view must have exactly one entry in its FROM list, which must be a table or another updatable view.
> 
>     The view definition must not contain WITH, DISTINCT, GROUP BY, HAVING, LIMIT, or OFFSET clauses at the top level.
> 
>     The view definition must not contain set operations (UNION, INTERSECT or EXCEPT) at the top level.
> 
>     The view's select list must not contain any aggregates, window functions or set-returning functions.


This also applies to the view you suggested in this thread:
`create view building_with_owners as select b.*, count(b.id) from qgisdbtest.building b JOIN qgisdbtest.building_owners o on o.id_building = b.id GROUP BY b.id;`

My own view behaves that way, so i need the ("invisible") original layer loaded as duplicate to be able to edit anything. It might be possible to work with` INSTEAD OF` - triggers, but at this point "straight forward" does not hold anymore for this solution imo.

I get that this is not a fast & easy to solve thing and requires time - i can start working on it  if you want and open a draft PR? Should we continue the discussion on https://github.com/qgis/QGIS/issues/61804 so we do not stay in an already closed issue?
 Also i might be able to organize some funding (for reviewing or implementing), if that is a way you'd be okay to move forward on?

**Comment by meyerlor** [comment]
I played around using claude, and asked it to patch it as it suggested. Following benchmarks were the result:

```
run | unpatched (master) | patched
-- | -- | --
1 | 151,443 ms | 3,208 ms
2 | 150,366 ms | 2,318 ms
3 | 152,031 ms | 2,328 ms
mean | 151,280 ms | 2,618 ms

```

Don't worry, I won't open a PR from that, it's too much AI generated code and i can not confidently say i understand what exactly was done. I see it more as a proof of concept.
